### ISSUE TITLE
dotnet publish - Razor page is found but returns blank response

### ISSUE THREAD

**Issue by glenn2223** [dotnet/aspnetcore#30947]
My issue relates to SDK's 5.0.103 & 5.0.201, however it could be the same with any previous SDK's

Upon publishing my site if a certain set of pages aren't loaded within the first few minutes of deployment then, when they are loaded, they will return a 200 with a blank screen/response.

~~To combat this I have to ping each of the problem URL's in a custom `lifetime.ApplicationStarted.Register` event~~ See [this comment](https://github.com/dotnet/aspnetcore/issues/30947#issuecomment-800093848)

To publish the site I use a PowerShell file. Within it, it runs `dotnet publish $projectPath -c Release -o $newPath -p:EnvironmentName=$env`, changes IIS's path to the `$newPath` and some other unimportant stuff

Please let me know what other information you require. Thanks in advance

**Comment by glenn2223** [comment]
Thanks for moving this over for me

---

I've included two screengrabs to hopefully assist in some way.

The first one is of the blank 200 response
![Blank 200 Response - image](https://user-images.githubusercontent.com/33450392/111282180-a08e4880-8635-11eb-931d-677c687487aa.png)

This second is of the same page, but it's upon visiting it straight after republishing (command ref below)
>To publish the site I use a PowerShell file. Within it, it runs `dotnet publish $projectPath -c Release -o $newPath -p:EnvironmentName=$env`, changes IIS's path to the `$newPath` and some other unimportant stuff

![OK 200 Response - Image](https://user-images.githubusercontent.com/33450392/111282719-33c77e00-8636-11eb-9092-c7d2f726822a.png)

---

Thoughts/things I've tried:
- Maybe to do with authentication
    - *Nope, it happened on both unauth and auth razor pages*
- Just razor related, not MVC (*I know, the View is razor*)
    - *Nope, it affects MVC pages as well*
- Maybe to do with caching
    - *Nope, set caching to `Duration = 0, NoStore = true`, it had no effect*

My last thought is that maybe it's hiding some kind of Exception? But, at the same time, why doesn't it work but then does after republishing?

**Comment by javiercn** [comment]
@glenn2223 thanks for contacting us.

Have you tried turning the log levels to Debug and looking at the output in those scenarios? We have not seen the problem you are describing and that might go a long way to figure out what's going on.

**Comment by glenn2223** [comment]
Thanks. Obvious now you've said it 🤦‍♂️

So now I've done that I'm getting the below, instead of a blank screen. Nothing in my DB logger though (obviously)
![Image of a 500.30 error screen](https://user-images.githubusercontent.com/33450392/111298990-4eeeb980-8647-11eb-99c7-c8b66196e33d.png)

**Comment by glenn2223** [comment]
So, upon javiercn's suggestion, I changed the logging from `Warning` to `Debug`. Upon publishing this change the site was producing 500 errors, as it couldn't start. I tracked it down to 2 issues. Both related to my custom logger, yet the first not *completely* my fault 😅.

#### TLDR;
1) EventId.Name is `string` and not `string?`, inserting this into a `NOT NULL` DB column threw
2) The were some concurrency issues with my custom `LoggerProvider`

Still strange that it didn't show a 500 error. My error page wouldn't have been shown due to some logging within it.

----

#### Longwinded version

So my custom logger inserts into a custom database table on each log. I have a custom function, because of this custom table, that converts a traditional log to my own usable logging method. Code copied below for ref
```c#
public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
{
    if (IsEnabled(logLevel))
        Log(logLevel, eventId.Name, state?.GetType().Name, "Log Triggered (Dependant)", formatter(state, exception), exception);
}
```
As you can see, the `Log(args)` method is passed `eventId.Name`. Code insight says it's a `string` but should be `string?` (I have nullability enabled, obv.). Upon inserting, further down the chain, it threw as I was inserting `NULL` into a `NOT NULL` column in my table.

My custom LoggerProvider, which I built using `services.BuildServices();` (naughty, I know), used the same instance of my `DbContext` for each of my custom loggers (very naughty, I know). This was fine at `Error`/`Critical` level, but once some serious logging starts (at `Debug`, for example) it's just not designed for it. I had planned to change it, based on [this here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0#no-asynchronous-logger-methods), but I hadn't got around to it yet.

So it looks like there were exceptions (most likely caused by these two points), that were hidden with no 500 error nor error page shown (as this did some logging too).

**Solutions:**
To fix point 1, I changed from `..., eventId.Name, ...` to `..., eventId.Name ?? eventId.Id.ToString(), ...` with an extra catch of nulls (this and possibly others) further down the chain.

To fix point 2, I finally transitioned my custom LoggerProvider from EFCore `SaveChanges()` to a shared in-memory `ConcurrentQueue` where the logs are dequeued and then transitioned to the DB within a background process (using, in this case, Hangfire).

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. Looks like this was caused by issues in your code. Glad that you've solved those.
If I'm missing something please let us know.

**Comment by glenn2223** [comment]
>1. EventId.Name is `string` and not `string?`, ...

Yes, but this may need addressing
